torchtune/utils/ppo_utils/_generation.py

```diff
@@ -0,0 +1,182 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Optional
+
+import torch
+from torchtune.utils._generation import sample, update_stop_tokens_tracker
+
+
+def generate_next_token(
+    model: torch.nn.Module,  # TODO (SalmanMohammadi) leaving as nn.module until TransformerDecoder refactor
+    input_pos: torch.Tensor,
+    x: torch.Tensor,
+    *,
+    mask: Optional[torch.Tensor] = None,
+    temperature: float = 1.0,
+    top_k: int = None,
+) -> torch.Tensor:
+    """Generates the next tokens."""
+    # model produces logits in [bsz, seq_length, vocab_size]
+    # we want to take the last token's logits as the input to the next model call
+    logits = model(x, input_pos=input_pos, mask=mask)[:, -1]
+    return sample(logits, temperature, top_k)
+
+
+def generate_next_token_with_value_head_model(
+    model: torch.nn.Module,  # TODO (SalmanMohammadi) leaving as nn.module until TransformerDecoder refactor
+    input_pos: torch.Tensor,
+    x: torch.Tensor,
+    *,
+    mask: Optional[torch.Tensor] = None,
+    temperature: float = 1.0,
+    top_k: int = None,
+) -> torch.Tensor:
+    """Generates the next tokens."""
+    # model produces logits in [bsz, seq_length, vocab_size]
+    # we want to take the last token's logits as the input to the next model call
+    logits, _ = model(x, input_pos=input_pos, mask=mask)
+    return sample(logits[:, -1], temperature, top_k)
+
+
+def get_causal_mask(
+    padding_mask: torch.Tensor,
+) -> torch.Tensor:
+    """
+    Converts an attention mask of shape `[bsz, seq_len]` to a causal attention mask suitable for
+    consumption by :func:`~torch.nn.functional.scaled_dot_product_attention~`.
+
+    HF uses a similar implementation internally, see
+    https://github.com/huggingface/transformers/blob/a564d10afe1a78c31934f0492422700f61a0ffc0/src/transformers/models/mistral/modeling_mistral.py#L1096
+
+    Args:
+        padding_mask (torch.Tensor): Boolean tensor where True indicates participation in attention
+            with shape [bsz x seq_length]
+    Returns:
+        torch.Tensor: Boolean causal mask with shape [bsz x seq_length x seq_length]
+    """
+    _, seq_len = padding_mask.shape
+    mask = torch.tril(
+        torch.ones(seq_len, seq_len, device=padding_mask.device, dtype=bool), diagonal=0
+    )
+    mask = mask & (padding_mask[:, None, :] & padding_mask[:, :, None])
+    mask.diagonal(dim1=1, dim2=2)[:] = True
+    return mask
+
+
+@torch.inference_mode()
+def generate(
```

Comment:
+ 1

Reply:
Hmmm, I had a quick try and looks like you're right. Was this added recently at all? I feel like when I tried it out I couldn't get it to work because generation required KV caches to be setup, and the causal mask used when setting up cacheing didn't account for padding? Maybe I'm misremembering.

Reply:
Update: this function now also returns logits and accepts a `torch.Generator`.